pkg/operations/k8sapi_ops_test.go

```diff
@@ -0,0 +1,239 @@
+		Expect(nodes[1].Name).To(Equal("k8s-agentpool3-1234"))
+		Expect(nodes[1].Status.Conditions[0].Type).To(Equal(v1.NodeOutOfDisk))
+		Expect(nodes[1].Status.Conditions[0].Status).To(Equal(v1.ConditionTrue))
+		Expect(nodes[1].Status.NodeInfo.KubeletVersion).To(Equal("1.9.9"))
+
+		// waiting for more nodes than the API returns should timeout
+		nodes, err = GetNodes(&mockClient, logger, apiserverURL, kubeconfig, timeout, "", 3)
+		var mostRecentGetNodesErr error
+
+		Expect(err).NotTo(BeNil())
+		Expect(err.Error()).To(Equal(fmt.Sprintf("GetAllNodes timed out: %s\n", mostRecentGetNodesErr)))
+		Expect(nodes).To(BeNil())
+
+		// waiting for fewer nodes than the API returns should timeout
+		nodes, err = GetNodes(&mockClient, logger, apiserverURL, kubeconfig, timeout, "", 1)
+
+		Expect(err).NotTo(BeNil())
+		Expect(err.Error()).To(Equal(fmt.Sprintf("GetAllNodes timed out: %s\n", mostRecentGetNodesErr)))
+		Expect(nodes).To(BeNil())
+
+		// filtering by pool name and and the waiting for the expected node count
+		nodes, err = GetNodes(&mockClient, logger, apiserverURL, kubeconfig, timeout, "agentpool3", 1)
+
+		Expect(err).To(BeNil())
+		Expect(nodes).To(HaveLen(1))
+		Expect(nodes[0].Name).To(Equal("k8s-agentpool3-1234"))
+		Expect(nodes[0].Status.Conditions[0].Type).To(Equal(v1.NodeOutOfDisk))
+		Expect(nodes[0].Status.Conditions[0].Status).To(Equal(v1.ConditionTrue))
+		Expect(nodes[0].Status.NodeInfo.KubeletVersion).To(Equal("1.9.9"))
+	})
+
+	It("GetNodes should return a meaningful timeout error when the k8s API fails to list nodes", func() {
+		mockClient := armhelpers.MockAKSEngineClient{MockKubernetesClient: &armhelpers.MockKubernetesClient{
+			FailListNodes: true,
+		}}
+		logger := log.NewEntry(log.New())
+		apiserverURL := "https://apiserver"
+		kubeconfig := "kubeconfig"
+		timeout := time.Second * 1 // set the timeout value high enough to allow for a single attempt
+
+		nodes, err := GetNodes(&mockClient, logger, apiserverURL, kubeconfig, timeout, "", -1)
+		mostRecentGetNodesErr := errors.New("ListNodes failed")
+
+		Expect(err).NotTo(BeNil())
+		Expect(err.Error()).To(Equal(fmt.Sprintf("GetAllNodes timed out: %s\n", mostRecentGetNodesErr)))
+		Expect(nodes).To(BeNil())
+	})
+
+	It("GetNodes should return a vanilla timeout error if timeout occurs before a single request", func() {
+		mockClient := armhelpers.MockAKSEngineClient{MockKubernetesClient: &armhelpers.MockKubernetesClient{}}
+		logger := log.NewEntry(log.New())
+		apiserverURL := "https://apiserver"
+		kubeconfig := "kubeconfig"
+		timeout := time.Second * 0 // by setting the timeout to 0 we time out immediately
+
+		nodes, err := GetNodes(&mockClient, logger, apiserverURL, kubeconfig, timeout, "", -1)
+		var mostRecentGetNodesErr error
+
+		Expect(err).NotTo(BeNil())
+		Expect(err.Error()).To(Equal(fmt.Sprintf("GetAllNodes timed out: %s\n", mostRecentGetNodesErr)))
+		Expect(nodes).To(BeNil())
+	})
+})
+
+func ExamplePrintNodes() {
```

Comment:
@CecileRobertMichon @mboersma FYI, in case you haven't seen this `Example<FuncName>` usage before: it's the golang idiom for validating stdout.

https://golang.org/pkg/testing/#hdr-Examples

Reply:
Hadn't seen it used in the wild yet. Achievement unlocked 🎖